src/clout/core.cljc

```diff
@@ -0,0 +1,150 @@
+(ns clout.core
+  "A small language for routing."
+  (:require [clojure.string :as string]
+            [clojure.set :as set]
+            [instaparse.core :as insta]))
+
+(def ^:private re-chars (set "\\.*+|?()[]{}$^"))
+
+(defn- re-escape [s]
+  #?(:clj
+     (string/escape s #(if (re-chars %) (str \\ %)))
+     :cljs
+     (let [keys (seq re-chars) vals (->> keys (map #(str "\\" %)))
+           cmap (zipmap keys vals)]
+       (string/escape s cmap))))
+
+(defn- re-groups*
+  #?@(:clj
+      [[^java.util.regex.Matcher matcher]
+       (for [i (range (.groupCount matcher))]
+         (.group matcher (int (inc i))))]
+      :cljs
+      [[re s]
+       (-> (re-seq re s) flatten rest)]))
+
+(defn- assoc-conj [m k v]
+  (assoc m k
+           (if-let [cur (get m k)]
+             (if (vector? cur)
+               (conj cur v)
+               [cur v])
+             v)))
+
+(defn- assoc-keys-with-groups [groups keys]
+  (reduce (fn [m [k v]] (assoc-conj m k v))
+          {}
+          (map vector keys groups)))
+
+(defn- request-url [request]
+  (str (name (:scheme request))
+       "://"
+       (get-in request [:headers "host"])
+       (:uri request)))
+
+(defn- path-info [request]
+  (or (:path-info request)
+      (:uri request)))
+
+(defprotocol Route
+  (route-matches [route request]
+    "If the route matches the supplied request, the matched keywords are
+    returned as a map. Otherwise, nil is returned."))
+
+(defrecord CompiledRoute [source re keys absolute?]
+  Route
+  (route-matches [_ request]
+    (let [path-info (if absolute?
+                      (request-url request)
+                      (path-info request))]
+      #?(:clj
+         (let [matcher (re-matcher re path-info)]
+           (if (.matches matcher)
+             (assoc-keys-with-groups (re-groups* matcher) keys)))
+         :cljs
+         (if (.test re path-info)
+           (assoc-keys-with-groups (re-groups* re path-info) keys)))))
+  #?@(:clj [Object
+            (toString [_] source)]))
+
+(def ^:private route-parser
+  (insta/parser
+    #?(:clj
+       "route    = (scheme / part) part*
```

Comment:
Can we avoid some reptition with `str`? I think I mentioned this in a previous review comment.

Reply:
Sorry, missed that.